### ISSUE TITLE
Size Mismatch in AdaptiveSoftmax when targets are not specified

### ISSUE THREAD

**Issue by amrsharaf** [pytorch/fairseq#213]
Following up on #212 , I'm updating `sequence_generator.py` to generate text from a pre-trained language model (initially trained with adaptive softmax). When computing log probabilities, and the targets are set to none, I receive a size mismatch exception in the line below, possibly because the dictionary size is smaller than the adaptive softmax cut-off: 

https://github.com/pytorch/fairseq/blob/388c520be21752cacb9fe3b1712038f32e0e9a5f/fairseq/modules/adaptive_softmax.py#L126

I imagine this could be solved by some sort of truncation to the output of tail[i].input

**Comment by huihuifan** [comment]
Why is the dictionary size smaller than the cutoff for adaptive softmax? Isn't the adaptive softmax cutoff's last bin value the size of the dictionary?

**Comment by amrsharaf** [comment]
I'm checking now, but I don't this this is the case, the last cut-off is an upper bound on the vocabulary size: https://github.com/pytorch/fairseq/blob/388c520be21752cacb9fe3b1712038f32e0e9a5f/fairseq/modules/adaptive_softmax.py#L23

**Comment by alexeib** [comment]
@amrsharaf how come your dictionary is smaller than the adaptive softmax? its true that if you specify adaptive softmax cutoff of 500,000 and your vocab size is 5,000 then you will get an adaptive softmax of size 500k, where only the first 5k is useful. you can work around that by taking the first ```vocab_size``` columns from what is returned by adaptive softmax, but in general the idea was that you should create adaptive softmax cutoffs that are compatible with your vocab size. we can put some assert to make sure that vocab size is at least as big as the last specified cutoff

**Comment by amrsharaf** [comment]
That's true, I should have made the max cut-off == vocab size, I just used the same hyperparameters from the wikitext README tutorial, it'd be helpful to add the assert statement or just an if statement like this: 

`if vocab_size < cutoff[i]: cutoff[i] = vocab_size`
